### PR TITLE
Update Guidelines

### PR DESCRIPTION
**Pull request by lesv**
* Update Contributing to closer match the current cononical version.
* Add PULL_REQUEST_TEMPLATE
* Update README to closer add the License requirement from the cononical version.
* Major revision of SAMPLE_FORMAT

### CODE REVIEW

File: SAMPLE_FORMAT.md [hunk 1]
```diff
@@ -25,19 +57,128 @@ specifically mentioned in the format, samples should make best-effort attempts i
...
+Snippets should have integration tests that should verify the snippet works and compiles correctly.
+Creating mocks for these tests are optional. These tests should capture output created by the
+snippet to verify that it works correctly. See the tests in the canonical for an example of how to
+do this correctly. 
+
+* Test Library: [JUnit4](https://junit.org/junit4/)
+* Test Runner: [Maven Failsafe plugin](https://maven.apache.org/surefire/maven-failsafe-plugin/)
+and [Maven Surefire plugin](https://maven.apache.org/surefire/maven-surefire-plugin/).
 
+As an example, the following test code shows how we test a region tag called `region_tag`:
```
Comment (ace-n) [inline SAMPLE_FORMAT.md]: Is it worth explaining what region tags are here? (Perhaps for external users and/or Googlers outside our team.)

File: SAMPLE_FORMAT.md [hunk 2]
```diff
@@ -25,19 +57,128 @@ specifically mentioned in the format, samples should make best-effort attempts i
...
+  in adding keys and secrets to the testing project.
+  
+### Running Locally
+Run tests locally with commands:
+
+* Maven: `mvn verify`
+* Gradle: `gradle build test`
+
+### Gradle Specifics
+#### NEEDS WORK
```
Comment (averikitsch) [inline SAMPLE_FORMAT.md]: Will this be updated in another PR?

Comment (lesv) [reply SAMPLE_FORMAT.md]: Yea - but we need to agree on a canonical build file, and add testing.  I didn't want to remove it, so I added that info.

File: SAMPLE_FORMAT.md [hunk 3]
```diff
@@ -75,50 +218,90 @@ should be used instead of explicitly declaring the client version. See the below
...
+To run the checkstyle & ErrorProne plugins on an existing sample, run
+
+```shell
+mvn clean verify -DskipTests
+```
+
+The `-DskipTests` is optional. It is useful if you want to verify that your code
+builds and adheres to the style guide without waiting for tests to complete.
+
+### Command-Line Arguments
```
Comment (averikitsch) [inline SAMPLE_FORMAT.md]: Should we note that this isn't recommended anymore?

File: SAMPLE_FORMAT.md [hunk 4]
```diff
@@ -25,19 +57,128 @@ specifically mentioned in the format, samples should make best-effort attempts i
...
+@RunWith(JUnit4.class)
+public class SomeClassTest {
+  @Test public void regionTag_shouldPass() throws Exception { }
+
+  @Test public void regionTag_shouldFail() throws Exception {
+    Assert.fail("should fail");
+  }
+}
+```
+You will note that the "_" in `region_tag` is removed, and "_" is used to separate regionTags from
```
Comment (ace-n) [inline SAMPLE_FORMAT.md]: This should read:
- underscores (`_`) are used in test method names to separate blocks of `camelCase` text ~from test descriptions~
- these blocks denote _region tags_, which serve as unique IDs for snippets in a given repository
- `camelCase` blocks beginning with `should` or `does` are **not** region tags - instead, they simply describe the test being run

File: SAMPLE_FORMAT.md [hunk 5]
```diff
@@ -75,50 +218,90 @@ should be used instead of explicitly declaring the client version. See the below
...
   duplicate instructions and README maintenance in the future. 
 
+# Code
+### Arrange, Act, Assert
+Samples should generally follow the "Arrange, Act, Assert" outline to: 
+* _Arrange_ - Create and configure the components for the request. Avoid nesting these components,
+  as complex, nested builders can be hard to read.
+* _Act_ - Send the request and receive the response.
+* _Assert_ - Verify the call was successful or that the response is correct. This is often done by
+  print contents of the response to `stdout`.
```
Comment (ace-n) [inline SAMPLE_FORMAT.md]: Nit: print**ing the**

File: SAMPLE_FORMAT.md [hunk 6]
```diff
@@ -1,15 +1,47 @@
...
+    * [Streams](#streams)
+    * [Parallel Streams](#parallel-streams)
+  * [Additional Best Practices](#additional-best-practices)
+    * [Time](#time)
+    * [Logging](#logging)
 
 This doc maintains an outline for 'snippet' samples specific to Java. Currently, the java canonical
 samples in this format are located 
-[here](https://github.com/GoogleCloudPlatform/java-docs-samples/tree/master/dlp/src/main/java/dlp/snippets).
+[here](/tree/master/dlp/src/main/java/dlp/snippets).
```
Comment (nnegrey) [inline SAMPLE_FORMAT.md]: Pretty sure this is just cus this is on a branch, but the link doesn't work.

File: SAMPLE_FORMAT.md [hunk 7]
```diff
@@ -131,7 +314,6 @@ they are example values only.
 Snippet functions should specific a return type of `void` and avoid returning any value wherever
```
Comment (nnegrey) [inline SAMPLE_FORMAT.md]:
```suggestion
Snippet functions should specify a return type of `void` and avoid returning any value wherever
```

File: SAMPLE_FORMAT.md [hunk 8]
```diff
@@ -211,21 +374,17 @@ Streams can be extremely compact, efficient, and easy to use - consider using th
...
-[Google Java Style Guide](https://google.github.io/styleguide/javaguide.html). It's encouraged, but
-not required to use `[google-java-format](https://github.com/google/google-java-format)` to help 
-format your code. 
-
 ### Time
 Use the `java.time` package when dealing with units of time in some manner. 
 
 ### Logging
-Use `java.util.logging` for consistent logging in web applications. 
+Use `java.util.logging` for consistent logging in web applications. Unless you are demonstrating how
```
Comment (kurtisvg) [inline SAMPLE_FORMAT.md]: Our [java logging examples](https://cloud.google.com/logging/docs/setup/java#example) tell users to use slf4j. Should we be consistent and start using that be default? (I believe it's compatible with Stackdriver too?)

File: SAMPLE_FORMAT.md (hunk 6 above)
Comment (lesv) [reply SAMPLE_FORMAT.md]: If there is a problem, I'll fix it w/ trivial PR.

File: SAMPLE_FORMAT.md (hunk 1 above)
Comment (lesv) [reply SAMPLE_FORMAT.md]: I'll put something in, but generally no.

File: SAMPLE_FORMAT.md [hunk 9]
```diff
@@ -75,50 +218,90 @@ should be used instead of explicitly declaring the client version. See the below
...
+which has optional parameters should use the [Apache Commons
+CLI](https://commons.apache.org/proper/commons-cli/index.html) library.
+
+**Dataflow** samples are an exception to this guideline, since Dataflow has its own 
+[method for setting options](https://cloud.google.com/dataflow/pipelines/specifying-exec-params).
+
+### Package Names
+It is not necessary to use a package for snippets.  If you choose to use one, please DO NOT use 
+`com.google.` as a prefix.  `com.example` or something related to the technology such as
+`package dlp.snippets`.
```
[the remark this answers is not in the document]
Comment (lesv) [reply SAMPLE_FORMAT.md]: Snippets are exactly what the default package was created for.

"The default package is an unnamed package. The unnamed package contains java classes whose source files did not contain a package declaration. The purpose of default package is for convenience when developing small or temporary applications or when just beginning development."

[the remark this answers is not in the document]
Comment (kurtisvg) [reply SAMPLE_FORMAT.md]: You shouldn't use it for production code, and users may not understand the subtlety and could take our endorsement of it that it's okay to use in other places. 

 https://stackoverflow.com/questions/7849421/is-the-use-of-javas-default-package-a-bad-practice

File: SAMPLE_FORMAT.md (hunk 8 above)
Comment (lesv) [reply SAMPLE_FORMAT.md]: Sg

**Comment by lesv** [comment]
@kurtisvg @averikitsch PTAL

File: SAMPLE_FORMAT.md [hunk 10]
```diff
@@ -75,53 +239,95 @@ should be used instead of explicitly declaring the client version. See the below
...
+builds and adheres to the style guide without waiting for tests to complete.
+
+### Command-Line Arguments
+**NOTE:** Snippet should be optimized to run directly from a user's IDE. Command-Line arguments are
+strongly discouraged, and new samples should not implement them. 
+
+Simple command-line samples with only positional arguments should use the
+`args` argument to `main(String... args)` directly. A command-line sample
+which has optional parameters should use the [Apache Commons
+CLI](https://commons.apache.org/proper/commons-cli/index.html) library.
```
Comment (kurtisvg) [inline SAMPLE_FORMAT.md]: Can we nix this part since we don't want folks to use CLI?

File: SAMPLE_FORMAT.md [hunk 11]
```diff
@@ -75,53 +239,95 @@ should be used instead of explicitly declaring the client version. See the below
...
+
+**Dataflow** samples are an exception to this guideline, since Dataflow has its own 
+[method for setting options](https://cloud.google.com/dataflow/pipelines/specifying-exec-params).
+
+### Package Names
+Samples should use either the default package or package names in the following formats: 
+`com.example.<YOUR_PRODUCT>.<FOLDER>`, such as `com.example.dlp.snippets` or 
+`com.example.functions.snippets`. Do not use `com.google.*`.
+
```
Comment (kurtisvg) [inline SAMPLE_FORMAT.md]: IMO, still should nix the default package.

**Comment by lesv** [comment]
PTAL

File: SAMPLE_FORMAT.md [hunk 12]
```diff
@@ -25,19 +57,149 @@ specifically mentioned in the format, samples should make best-effort attempts i
...
+```
+You will note:
+ * underscores (`_`) are used in test method names to separate blocks of `camelCase` text
+ * these blocks denote _region tags_, which serve as unique IDs for snippets in a given repository
+ * `camelCase` blocks beginning with `should` or `does` are **not** region tags - instead, they
+ simply describe the test being run
+
+
+that the "_" in `region_tag` is removed, and "_" is used to separate regionTags from
+test descriptions.
```
Comment (dzlier-gcp) [inline SAMPLE_FORMAT.md]: Incomplete?

File: SAMPLE_FORMAT.md [hunk 13]
```diff
@@ -25,19 +57,149 @@ specifically mentioned in the format, samples should make best-effort attempts i
...
+please contact one of the repo admins.
+
+### Testing Setup
+Most samples require a GCP project and billing account. Keep the following in
+mind when setting up tests.
+
+* **Environment variables**  
+  Minimize additional environment variables that need to be set to run the tests.
+  If you do require additional environment variables, they should be added to
+  `run_tests.sh`.
```
Comment (dzlier-gcp) [inline SAMPLE_FORMAT.md]: Link to file?

File: SAMPLE_FORMAT.md [hunk 14]
```diff
@@ -25,19 +57,149 @@ specifically mentioned in the format, samples should make best-effort attempts i
...
+  If an API needs to be enabled in the testing project, add this information to the
+  directory's README and to the comments in the PR. If there is no README.md file, add one
+  in your PR.
+
+* **IAM**
+  Some API's require that the service account have some additional capibilities, these should also
+  be mentioned in both the PR and the README.
+
+* **Cloud resources**  
+  Most Java samples create the Cloud resources that they need to run. If this
```
Comment (dzlier-gcp) [inline SAMPLE_FORMAT.md]: As per the "keeping tests from getting stale" lightning talk we recently had, might want to add a note about how tests that create cloud resources should also delete those resources when they are done testing, particularly in a way that will ensure the deletion of the resource even if the test fails, such as with with a `finally` block or in an `@After` or `@AfterClass` function.

File: SAMPLE_FORMAT.md [hunk 15]
```diff
@@ -211,21 +394,17 @@ Streams can be extremely compact, efficient, and easy to use - consider using th
 * Checked Exceptions can be problematic inside streams.
   
 ### Parallel Streams
-Parallel Streams make make sense in a few situations. There are many situations where there use is a net loss.  Really think through your usage and consider what they might mean if you are already doing concurrent operations.
+Parallel Streams make make sense in a few situations. There are many situations where there use is a
```
Comment (dzlier-gcp) [inline SAMPLE_FORMAT.md]: make make sense

File: SAMPLE_FORMAT.md (hunk 12 above)
Comment (lesv) [reply SAMPLE_FORMAT.md]: Extra -- good catch - thanks.

File: SAMPLE_FORMAT.md (hunk 14 above)
Comment (lesv) [reply SAMPLE_FORMAT.md]: done

File: SAMPLE_FORMAT.md (hunk 13 above)
Comment (lesv) [reply SAMPLE_FORMAT.md]: done

**Comment by lesv** [comment]
A couple more changes if anyone want's one last look at it before it automerges.
